extra/generate_option_list.cc

```diff
@@ -0,0 +1,359 @@
+  while (*command)
+  {
+    switch (*command) {
+    case '\"':
+    case '\\':
+    case '$':
+      escaped_command.push_back('\\');
+      break;
+    default:
+      escaped_command.push_back(*command);
+    }
+    command++;
+  }
+  escaped_command.push_back('\"');
+  return escaped_command;
+}
+
+std::string read_output(FILE *f)
+{
+  char buf[4096];
+  size_t read_bytes;
+  std::string output;
+  do
+  {
+    read_bytes= std::fread(buf, 1, sizeof buf, f);
+    output.append(buf, read_bytes);
+  } while (read_bytes > 0);
+  return output;
+}
+
+std::string call_mariadbd(const char *mariadbd_path)
+{
+  std::string output;
+  std::string command= escape_command(mariadbd_path) +
+                       " --no-defaults"
+                       " --plugin-maturity=unknown"
+                       " --plugin-load=\""
+                       "adt_null;"
+                       "auth_0x0100;"
+                       "auth_ed25519;"
+                       "auth_gssapi;"
+                       "auth_pam;"
+                       // "auth_pam_v1;"
+                       "auth_test_plugin;"
+                       "cracklib_password_check;"
+                       "debug_key_management;"
+                       "dialog_examples;"
+                       "disks;"
+                       "example_key_management;"
+                       "file_key_management;"
+                       "func_test;"
+                       "ha_archive;"
+                       "ha_blackhole;"
+                       "ha_connect;"
+                       "ha_federatedx;"
+                       "ha_mroonga;"
+                       "handlersocket;"
+                       // "ha_oqgraph;"
+                       "ha_rocksdb;"
+                       "ha_s3;"
+                       "hashicorp_key_management;"
+                       "ha_sphinx;"
+                       "ha_spider;"
+                       "ha_test_sql_discovery;"
+                       "libdaemon_example;"
+                       "locales;"
+                       "metadata_lock_info;"
+                       "mypluglib;"
+                       "password_reuse_check;"
+                       "provider_bzip2;"
+                       "provider_lz4;"
+                       "provider_lzma;"
+                       "qa_auth_interface;"
+                       "qa_auth_server;"
+                       "query_cache_info;"
+                       "query_response_time;"
+                       "server_audit;"
+                       "simple_password_check;"
+                       "sql_errlog;"
+                       "test_sql_service;"
+                       "test_versioning;"
+                       "type_mysql_json;"
+                       "type_mysql_timestamp;"
+                       "type_test;"
+                       "wsrep_info"
+                       "\""
+                       " --verbose"
+                       " --help";
+
+  FILE *f= my_popen(command.c_str(), "r");
```

Comment:
`cmake -DWITH_UBSAN=ON` using GCC 13.2.0 does not give any hints. But, there is a promising [-DWITH_MSAN=ON build](https://buildbot.mariadb.org/#/builders/572/builds/6184):
```
Uninitialized bytes in __interceptor_popen at offset 2 inside [0x717000000000, 805)
==12326==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55a357b8085a in call_mariadbd(char const*) /home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list.cc:128:12
    #1 0x55a357b87f19 in main /home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list.cc:356:20
    #2 0x7f29c62a2d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x23d09) (BuildId: a3780b0b8a5bf5876e31d16b0a9d8fc6ba69a1f2)
    #3 0x55a357af8519 in _start (/home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list+0x25519) (BuildId: f17428483dcac03d445c4d10f154c049a08fd510)
  Uninitialized value was stored to memory at
    #0 0x55a357b25db4 in __msan_memmove (/home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list+0x52db4) (BuildId: f17428483dcac03d445c4d10f154c049a08fd510)
    #1 0x7f29c669c3ca in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(char const*, unsigned long) (/usr/lib/x86_64-linux-gnu/libc++.so.1+0x5b3ca) (BuildId: bdd6136d41123bda4a991a8ff4594c1c11f3d56d)
  Uninitialized value was stored to memory at
    #0 0x55a357b25db4 in __msan_memmove (/home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list+0x52db4) (BuildId: f17428483dcac03d445c4d10f154c049a08fd510)
    #1 0x7f29c669c818 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::push_back(char) (/usr/lib/x86_64-linux-gnu/libc++.so.1+0x5b818) (BuildId: bdd6136d41123bda4a991a8ff4594c1c11f3d56d)
  Uninitialized value was stored to memory at
    #0 0x55a357b25db4 in __msan_memmove (/home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list+0x52db4) (BuildId: f17428483dcac03d445c4d10f154c049a08fd510)
    #1 0x7f29c669c818 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::push_back(char) (/usr/lib/x86_64-linux-gnu/libc++.so.1+0x5b818) (BuildId: bdd6136d41123bda4a991a8ff4594c1c11f3d56d)
  Uninitialized value was created by an allocation of 'ref.tmp' in the stack frame of function '_Z13call_mariadbdPKc'
    #0 0x55a357b804a0 in call_mariadbd(char const*) /home/buildbot/amd64-debian-11-msan/build/extra/generate_option_list.cc:70
```
I will try to reproduce this.

Reply:
I can reproduce this locally on `clang-15`:
```
cd /dev/shm/11m-config/extra && /dev/shm/11m-config/extra/generate_option_list /dev/shm/11m-config/sql/mariadbd /dev/shm/11m-config/include/mariadbd_options.h
Uninitialized bytes in __interceptor_popen at offset 2 inside [0x717000000000, 783)
==1202847==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55db7b334831 in call_mariadbd(char const*) /mariadb/11-config/extra/generate_option_list.cc:128:12
    #1 0x55db7b33c381 in main /mariadb/11-config/extra/generate_option_list.cc:356:20
    #2 0x7f9ad746e6c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #3 0x7f9ad746e784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #4 0x55db7b2a9730 in _start (/dev/shm/11m-config/extra/generate_option_list+0x7d730) (BuildId: 9864fc16f8ed3435)

  Uninitialized value was stored to memory at
    #0 0x55db7b2d7ef8 in __msan_memmove (/dev/shm/11m-config/extra/generate_option_list+0xabef8) (BuildId: 9864fc16f8ed3435)
    #1 0x7f9ad76ff193 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(char const*, unsigned long) (/lib/x86_64-linux-gnu/libc++.so.1+0x59193) (BuildId: 9529825510c549922deaf45f996aaf182b1ecb4e)

  Uninitialized value was stored to memory at
    #0 0x55db7b2d7ef8 in __msan_memmove (/dev/shm/11m-config/extra/generate_option_list+0xabef8) (BuildId: 9864fc16f8ed3435)
    #1 0x7f9ad76ff5f2 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::push_back(char) (/lib/x86_64-linux-gnu/libc++.so.1+0x595f2) (BuildId: 9529825510c549922deaf45f996aaf182b1ecb4e)

  Uninitialized value was created by an allocation of 'ref.tmp' in the stack frame of function '_Z13call_mariadbdPKc'
    #0 0x55db7b334470 in call_mariadbd(char const*) /mariadb/11-config/extra/generate_option_list.cc:70
```

Reply:
If I got the shadow byte addresses right, there are 32 uninitialized bytes in the middle of the string:
```
(rr) p (char*)$rdi
$3 = 0x717000000000 "\"/dev/shm/11m-config/sql/mariadbd\" --no-defaults --plugin-maturity=unknown --plugin-load=\"adt_null;auth_0x0100;auth_ed25519;auth_gssapi;auth_pam;auth_test_plugin;cracklib_password_check;debug_key_mana"...
(rr) p *(char*)0x717000000000@782
$4 = "\"/dev/shm/11m-config/sql/mariadbd\" --no-defaults --plugin-maturity=unknown --plugin-load=\"adt_null;auth_0x0100;auth_ed25519;auth_gssapi;auth_pam;auth_test_plugin;cracklib_password_check;debug_key_mana"...
(rr) p *(char*)0x217000000000@782
$5 = "\000\000", '\377' <repeats 32 times>, '\000' <repeats 747 times>
```
That is, the `/dev/shm/11m-config/sql/mariadbd` is claimed to be uninitialized. It was marked as such here:
```
(rr) watch -l *(char*)0x217000000002@8
Hardware watchpoint 4: -location *(char*)0x217000000002@8
(rr) rc
Continuing.

Hardware watchpoint 4: -location *(char*)0x217000000002@8

Old value = "\377\377\377\377\377\377\377\377"
New value = "\000\000\000\000\000\000\000"
__memset_avx2_unaligned_erms () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:244
244	../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S: Tiedostoa tai hakemistoa ei ole.
(rr) bt
#0  __memset_avx2_unaligned_erms () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:244
#1  0x000055e5307ad7df in __msan::MsanAllocate(__sanitizer::StackTrace*, unsigned long, unsigned long, bool) ()
#2  0x000055e5307ad6fa in __msan::msan_malloc(unsigned long, __sanitizer::StackTrace*) ()
#3  0x000055e53080c149 in operator new(unsigned long) ()
#4  0x00007f33b696417c in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::append(char const*, unsigned long) () from /lib/x86_64-linux-gnu/libc++.so.1
#5  0x000055e53080d6e9 in std::__1::operator+[abi:v160006]<char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, char const*) (
    __lhs=@0x7ffdf6110d90: {__r_ = {<std::__1::__compressed_pair_elem<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__rep, 0, false>> = {__value_ = {{__l = {{__is_long_ = 1, __cap_ = 24}, __size_ = 34, __data_ = 0x703000000000 "\"/dev/shm/11m-config/sql/mariadbd\""}, __s = {{__is_long_ = 1 '\001', __size_ = 24 '\030'}, __padding_ = 0x7ffdf6110d91 "", __data_ = "\000\000\000\000\000\000\000\"", '\000' <repeats 11 times>, "0p\000"}, __r = {__words = {49, 34, 123351460741120}}}}}, <std::__1::__compressed_pair_elem<std::__1::allocator<char>, 1, true>> = {<std::__1::allocator<char>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<char> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}}, __rhs=0xff <error: Cannot access memory at address 0xff>)
    at /usr/include/c++/v1/string:4374
#6  call_mariadbd (mariadbd_path=<optimized out>) at /mariadb/11-config/extra/generate_option_list.cc:72
#7  0x000055e530815382 in main (argc=3, argv=0x7ffdf6111fd8) at /mariadb/11-config/extra/generate_option_list.cc:356
```
The problem appears to be the `escape_command()` call.

I was also able to reproduce this with `MSAN_OPTIONS=poison_in_dtor=0` ([MDEV-30942](https://jira.mariadb.org/browse/MDEV-30942)), which I suppose is set on buildbot anyway.

Reply:
That code was only copying already declared-as-uninitialized data. I single-stepped to the `memmove` call inside `std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::append(char const*, unsigned long)` and checked the shadow bytes of the source string (`$rsi` minus the shadow byte offset). But I did not really understand where this uninitialized data would come from.
I tried applying the following patch, with no improvement:
```diff
diff --git a/extra/generate_option_list.cc b/extra/generate_option_list.cc
index 1afd17f76cb..c557b76c89b 100644
--- a/extra/generate_option_list.cc
+++ b/extra/generate_option_list.cc
@@ -69,7 +69,8 @@ std::string read_output(FILE *f)
 std::string call_mariadbd(const char *mariadbd_path)
 {
   std::string output;
-  std::string command= escape_command(mariadbd_path) +
+  std::string command= escape_command(mariadbd_path);
+  command +=
                        " --no-defaults"
                        " --plugin-maturity=unknown"
                        " --plugin-load=\""
```
This led to the following output:
```sh
_RR_TRACE_DIR=/dev/shm/rr MSAN_OPTIONS=msan_track_origins=1 rr record /dev/shm/11m-config/extra/generate_option_list /dev/shm/11m-config/sql/mariadbd /dev/shm/11m-config/include/mariadbd_options.h
```
```
rr: Saving execution to trace directory `/dev/shm/rr/generate_option_list-6'.
Uninitialized bytes in __interceptor_popen at offset 2 inside [0x717000000000, 783)
==1207137==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x562466ffb725 in call_mariadbd(char const*) /mariadb/11-config/extra/generate_option_list.cc:129:12
    #1 0x5624670031c1 in main /mariadb/11-config/extra/generate_option_list.cc:357:20
    #2 0x7fc0e9bc06c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #3 0x7fc0e9bc0784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #4 0x562466f70730 in _start (/dev/shm/11m-config/extra/generate_option_list+0x7d730) (BuildId: ea4ce4857a1fd6a2)

  Uninitialized value was stored to memory at
    #0 0x562466f9eef8 in __msan_memmove (/dev/shm/11m-config/extra/generate_option_list+0xabef8) (BuildId: ea4ce4857a1fd6a2)
    #1 0x7fc0e9e51193 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::append(char const*, unsigned long) (/lib/x86_64-linux-gnu/libc++.so.1+0x59193) (BuildId: 9529825510c549922deaf45f996aaf182b1ecb4e)

  Uninitialized value was stored to memory at
    #0 0x562466f9eef8 in __msan_memmove (/dev/shm/11m-config/extra/generate_option_list+0xabef8) (BuildId: ea4ce4857a1fd6a2)
    #1 0x7fc0e9e515f2 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::push_back(char) (/lib/x86_64-linux-gnu/libc++.so.1+0x595f2) (BuildId: 9529825510c549922deaf45f996aaf182b1ecb4e)

  Uninitialized value was created by an allocation of 'command' in the stack frame of function '_Z13call_mariadbdPKc'
    #0 0x562466ffb470 in call_mariadbd(char const*) /mariadb/11-config/extra/generate_option_list.cc:70

SUMMARY: MemorySanitizer: use-of-uninitialized-value /mariadb/11-config/extra/generate_option_list.cc:129:12 in call_mariadbd(char const*)
Exiting
```
It could be safer to use `std::ostringstream`, which we use elsewhere.

Reply:
> It could be safer to use `std::ostringstream`, which we use elsewhere.

Isn't that just a ([slow](https://lemire.me/blog/2023/10/19/for-processing-strings-streams-in-c-can-be-slow/)) wrapper over strings? I just pushed a version that uses `ostringstream` anyway.

Reply:
Yeah, the problem still remains, but now we got some more detailed hint:
```
cd /dev/shm/11m-config/extra && /dev/shm/11m-config/extra/generate_option_list /dev/shm/11m-config/sql/mariadbd /dev/shm/11m-config/include/mariadbd_options.h
==1215051==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55d417c97f3d in std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) /usr/include/c++/v1/ostream:759:13
    #1 0x55d417c86fc8 in std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::operator<<[abi:v160006]<std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char) /usr/include/c++/v1/ostream:827:12
    #2 0x55d417c86fc8 in escape_command(std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, char const*) /mariadb/11-config/extra/generate_option_list.cc:39:7
    #3 0x55d417c87728 in call_mariadbd(char const*) /mariadb/11-config/extra/generate_option_list.cc:73:3
    #4 0x55d417c8f4a1 in main /mariadb/11-config/extra/generate_option_list.cc:357:20
    #5 0x7f9db520f6c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #6 0x7f9db520f784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #7 0x55d417bfc640 in _start (/dev/shm/11m-config/extra/generate_option_list+0x7e640) (BuildId: 5f5e24aa74e03b40)

  Uninitialized value was created by an allocation of '__s' in the stack frame of function '_ZNSt3__124__put_character_sequenceB7v160006IcNS_11char_traitsIcEEEERNS_13basic_ostreamIT_T0_EES7_PKS4_m'
    #0 0x55d417c979e0 in std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) /usr/include/c++/v1/ostream:753

SUMMARY: MemorySanitizer: use-of-uninitialized-value /usr/include/c++/v1/ostream:759:13 in std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long)
```
I got a bit better stack trace with
```sh
MSAN_OPTIONS=msan_track_origin=1:abort_on_error=1 _RR_TRACE_DIR=/dev/shm/rr rr record /dev/shm/11m-config/extra/generate_option_list /dev/shm/11m-config/sql/mariadbd /dev/shm/11m-config/include/mariadbd_options.h
rr replay /dev/shm/rr/latest-trace
```
like this:
```
(rr) set print frame-arguments scalar
(rr) bt
#0  std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) (__os=..., 
    __str=__str@entry=0x7ffddd28730d "\"", __len=__len@entry=1) at /usr/include/c++/v1/ostream:758
#1  0x0000557cfdc24fc9 in std::__1::operator<<[abi:v160006]<std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char) (__os=..., __c=34 '"')
    at /usr/include/c++/v1/ostream:827
#2  escape_command (out=..., command=command@entry=0x7ffddd28a86f "/dev/shm/11m-config/sql/mariadbd") at /mariadb/11-config/extra/generate_option_list.cc:39
#3  0x0000557cfdc25729 in call_mariadbd (mariadbd_path=0x7ffddd28a86f "/dev/shm/11m-config/sql/mariadbd") at /mariadb/11-config/extra/generate_option_list.cc:73
#4  0x0000557cfdc2d4a2 in main (argc=3, argv=0x7ffddd288758) at /mariadb/11-config/extra/generate_option_list.cc:357
```
All MSAN shadow bytes corresponding to `mariadbd_path` are 0, meaning that the string is initialized. Besides, there should be an error when branching on an uninitialized `*command` already. The amd64-centos-7 build also still fails.

Another idea is that we have a dependency on `libfmt`. It might be used here. Or just `my_snprintf` or some variant. It might implement some suitable escaping already.

Reply:
An old-fashioned way might be to determine the length in one call, then allocate a C `char` array, and finally copy the escaped output to the buffer. Too bad that neither `snprintf()` nor `my_snprintf()` support anything that would escape string literals.

Reply:
I finally figured it out. Sorry, I did not realize that `cmake --build .` (or whatever we use to build) will invoke this program without pointing `LD_LIBRARY_PATH` to a `clang -fsanitize=memory` instrumented `libc++`.

If I specify a suitable `LD_LIBRARY_PATH`, then the invocation will complete without errors. I am really sorry for this time waste for us both. But I guess the easiest way to work around this might be to use C strings and avoid any C++.

The crash on amd64-centos-7 must be something else, and you’d probably need remote access to the environment to debug that.

Reply:
We might also tweak the MSAN builder configuration so that a suitable `LD_LIBRARY_PATH` (pointing only to `libc++` and not the rest) will be set during the build already.

Reply:
There should be a practical work-around to get the MSAN build to work: Add the following right before the definition of the function that calls `my_popen`:
```c++
#if defined(__has_feature)
#  if __has_feature(memory_sanitizer)
     __attribute__((no_sanitize("memory"))) /* we may run with uninstrumented std::string */
#  endif
#endif
```
We have that in `slide_hash()` in `zlib/deflate.c`.

Reply:
I tried putting that to both `call_mariadbd` and `escape_command`, but [the MSAN CI is still failing](https://buildbot.mariadb.org/#/builders/572/builds/6217)

Reply:
There are a couple of alternative options:

1. Modify the `CMAKE_CXX_FLAGS` for this target by removing any `-fsanitize=memory -fsanitize-memory-track-origins`. Hopefully @vaintroub can help with that.
2. Change the tool invocation in CMake, by extending the definition of `LD_LIBRARY_PATH` with the contents of another environment variable if available. The invocation could be something like this:
```sh
GENERATOR_LD_LIBRARY_PATH=/path/to_msan_instrumented_libc++ cmake --build .
```
3. (hacky) No source code change, but change the compilation invocation to something like the following:
```sh
LD_LIBRARY_PATH=/path/to_msan_instrumented_libc++ cmake --build .
```
This is risky, because if anything else that is invoked during the build happens to be dynamically linked to `libc++.so` (`clang` could be a good candidate, although on my system it uses `libstdc++` instead), then those commands would fail to start due to an unresolved symbol because the MSAN-instrumented library would depend on extra symbols that are part of each MSAN instrumented executable.

Reply:
I do not know if 1) will work. It links with mysys, which is instrumented, and one can't remove the flags from mysys.
I'd say - just do not build it with MSAN, and make it an optional thing, also in mtr.

BTW, there are also problems with cross-compilation builds. One can't invoke TARGET_FILE:mariadbd in cross-compilation, that would not start.

Reply:
There is another hacky solution: Call `MEM_MAKE_DEFINED()` to "bless" the string after the non-inline function had copied the `mariadbd_path` to it. In any case, it might make sense to go back to the original `perl` script approach. This executable would be for maintainers only, for manually creating a list of options or option values that need to be replaced. The generated output would then be checked into the repository, a bit like the `flex` and `bison` generated files under `storage/innobase`. Not kosher, but maybe the least pain for something that is not expected to change often.